### DISCUSSION TITLE
Software Versioning Working Group Summit

### DISCUSSION THREAD

**Discussion by nieuwsma** [Cray-HPE/community#25] [CSM-Architecture SIG]
# Purpose

We'd like to hammer out some standard regarding how we version our applications and CSM releases.

## Attendees
The attendees are the CSM tech leads who are interested in the topic.

- @nieuwsma - Facilitator 
- @mpkelly-hpe 
- @zcrisler 
- @dborman-hpe 
- @jsollom-hpe 
- @rkleinman-hpe 
- @tpletcher-hpe 
- @SeanWallace 
- @denniswalker 

## Objectives

Develop a basic shared understanding for how to handle the majority of repositories (80%). We would like this understanding to not be 'tribal' but be codified, reasoned, and communcatable.   These are not laws, but what we are striving for; they are patterns to follow.

Goals:

- Provide transparency to CSM release management about the capability, requirements, compatibilities of our software.
- Provide clear, unambiguous, actionable processes for software development that do not hinder developer agility, velocity, or product maturity. _THIS IS EXPENSIVE IN REAL DEVELOPER TIME, and MAKES SOFTWARE BRITTLE_

Topics for consideration:

1. Clarify how 'release' branches are created in individual repos.  How do we escape long lived development branches?
2. How to give guidance/clarity to CSM manifest reviewers to recognize when features might be getting accidentally pulled into an incorrect release?  
3. How does CSM manifest reviewers understand what software is shipping?
4. develop terminology and standards for what level of testing is required for 'features' or code changes? e.g. is a Full CSM test? release cycle test?  
5. How to deal with saying "feature foo is slated for release x.y, and can't go to release x.(y-1)

RELATED: https://connect.us.cray.com/confluence/display/~dborman/CSM+Versioning+and+Release+Branching

**Comment by rkleinman-hpe** [comment]
I'd like to add to the topics:

1. Recommendations for changelog format/location/etc (and/or Github releases) in individual code repositories to aid integrators/release managers in determining how feature creep can be mitigated.
2. Recommendations for individual repositories on how to describe the repo's branching strategy (i.e. "we follow gitflow" in a README file).
3. How versioning and branching strategies will propagate to external contributors once CSM is fully open sourced to the public. External submissions should not be hindered by any process that is special or only accessible to HPE employees.

**Comment by rkleinman-hpe** [comment]
Another topic to discuss at the next meeting:

* How do we handle versioning with CVE that change base images? We have recently had far reaching CVEs for OpenSSL which affected a significant number of the base images our containers are built upon. In this instance, we aren't changing the HPE code, but we are forced to bump the versions to get a new build into the CSM manifests. Might this be an argument for container versions that include build metadata (timestamps, build numbers)?

For example, in our current model:

    v2.5.7 (contains CVE-affected base image)
    v2.5.8 (CVE mitigated)

Would it not be more correct (semantic versioning wise) to have:

    v2.5.7-202107280326 (contains CVE)
    v2.5.7-202109281642 (CVE mitigated)

Does updating the base image for one of our container images count as a .z patch (in semver lingo)?

**Reply by nieuwsma** [reply]
IMO I think if the fix for the CVE is just a 'rebuild' I think that is a great argument for build metadata.  If we actually have to change code then I think we should follow normal semver rules.   Does that make sense?

**Reply by jsollom-hpe** [reply]
I agree.

**Reply by rkleinman-hpe** [reply]
So with build metadata, it would be `v2.5.7+202109281642`?

**Reply by dborman-hpe** [reply]
So, is the v2.5.7 the version of the source code used to generate the container image, or the version of the container image?  If it is the version of the container image, then you bump the patch number when you rebuild, because it is a new container image.  But if you think the version is the version of the source code, then that could be an argument for adding build information.

**Reply by rkleinman-hpe** [reply]
> So, is the v2.5.7 the version of the source code used to generate the container image, or the version of the container image? If it is the version of the container image, then you bump the patch number when you rebuild, because it is a new container image. But if you think the version is the version of the source code, then that could be an argument for adding build information.

Hadn't thought about the code version being different than the container version. I believe in all cases for CMS repos, if a container is built, it is synonymous with the code version.

**Reply by dborman-hpe** [reply]
> > So, is the v2.5.7 the version of the source code used to generate the container image, or the version of the container image? If it is the version of the container image, then you bump the patch number when you rebuild, because it is a new container image. But if you think the version is the version of the source code, then that could be an argument for adding build information.
> 
> Hadn't thought about the code version being different than the container version. I believe in all cases for CMS repos, if a container is built, it is synonymous with the code version.

Yes, these are the nuances that can cause different views on when to or when not to change the version, based on different assumptions that the individuals themselves may not even be aware they are making!

**Comment by jsollom-hpe** [comment]
I have several topics for future discussion.

### Workflows
Do we have a workflow that we want to recommend to development teams? Martin Fowler references three different popular work-flows in his voluminous blog on branching patterns: https://martinfowler.com/articles/branching-patterns.html#PersonalThoughtsOnIntegrationPatterns

They are
1. Git-flow
2. Github-flow
3. Trunk-driven development (i.e. Continuous integration)

There may be other workflows, too. I feel git-flow likely most closely aligns with our need to maintain long-lived release branches. Martin favored continuous integration as the most effective and efficient from a developer stand-point. It may be cost-prohibitive to achieve a 'Healthy Branch' model to support continuous integration.

I feel that the choice of workflow can be unique to each repository. It can depend on the developers working in that repository and the overall health of the repository. 

I do not feel we can dictate a workflow, only recommend one.

### Versioning
As previous comments allude to, we need to work on versioning. 
We need versions to insert into the product manifest file to identify our Helm chart, Docker image, and rpms.
Should these versions be separate and independent? Should they be the same version because they are considered a 'cohesive unit'?

We have already decided to use SemVer for versioning. If we do not cut future-looking release branches, will we avoid having to do any artificial 'version bumping' to give ourselves headspace for later branches? I think that may depend upon the workflow chosen for a repository. We should figure that out. 

There are some basic questions to answer regarding versioning?
1. Do we have to increment the version for every commit? 
2. Do we need version files in our software? How many? One per Helm, Docker, and rpm? 
3. What are we using versions for? 
3.1 Debugging
3.2 Indicating software capabilities

In CMS, we have dispensed with a version per artifact and just given each the same version. Viewing them as one cohesive unit has simplified our lives, for the most part. We also had to come up with a slightly complicated versioning strategy to avoid file merge conflicts for the version files. We don't have hard-coded version files. They are calculated at build time based off of a delta of the number of build commits that  this commit is from a given base. We may not need a strategy like this any longer because we are no longer cutting future-looking release branches. However, we should organize our branches in such a way that we no longer need devices like this.

### Tagging
How does tagging correspond to versioning? Do version files and tags provide the same information? If not, how are they different? If they are the same, why do we have this redundancy?  

Are we going to use tags to trigger publishing artifacts to stable locations? I think we already are. 
As we discussed, do we need individual tags that correspond to the individual build artifacts? A tag for a Helm chart, one for a Docker file, one for an rpm. Those are now three tags to remember where before we had only one. Further, we would need to set up the build infrastructure to recognize and act on those tags.

### Release Branching
When do we want to cut release branches?
I would argue as little as possible. If we had a health mainline branch that we could always release from, then we could only cut a release branch if these two conditions are met.
1. There is a fix that needs to be made to a published/production piece of code.
2. There is something in the mainline branch that cannot be allowed to go to the customer. This would be a feature added to the mainline that was incompatible with the other software running on the system.

**Reply by rkleinman-hpe** [reply]
My opinions below:

### Re: Workflows

I think the Fowler article should be able to cover most, if not all, CASM use cases and we should use that as the official recommendation. It is so complete and provides options for several situations. In my opinion, CASM as a whole should never need or have to mandate any branching workflow on any repository. We already know and have experienced that a one-size-fits-all situation will fail given the diversity of our repositories. Given the nature of Shasta and CSM where we will have multiple versions of our software in the field, GitFlow is likely a good default candidate for many of our repositories.

### Re: Tagging

Again, I believe each repository should be free to provide artifacts as they see fit, whether that is building the head of a release branch or building commits based on tags. As long as the repository `README`/`Contributing.md` file is accessible and clear on the strategy of the repository, everyone should be able to build artifacts cleanly in a repeatable fashion.

If repositories follow GitFlow, tagging is the way that build artifacts are versioned on the master/main branch. From [GitFlow](https://nvie.com/posts/a-successful-git-branching-model/#the-main-branches):

> Therefore, each time when changes are merged back into master, this is a new production release by definition. We tend to be very strict at this, so that theoretically, we could use a Git hook script to automatically build and roll-out our software to our production servers every time there was a commit on master.

If a tag provides the version, there is no reason for version files. Builds can grab tag data to fill in files that need version information. Many repos do this already.

### Re: Release Branching

When to cut release branches is a repo-by-repo decision. Some repos may decide that they never need a release branch. In [GitFlow](https://nvie.com/posts/a-successful-git-branching-model/#release-branches) they are done only when the release is ready or almost ready:

> Release branches support preparation of a new production release. They allow for last-minute dotting of i’s and crossing t’s. Furthermore, they allow for minor bug fixes and preparing meta-data for a release (version number, build dates, etc.).
